### PR TITLE
Add deadlock detection

### PR DESCRIPTION
**Pull request by Zoxc**

### CODE REVIEW

File: rayon-core/src/sleep/mod.rs [hunk 1]
```diff
@@ -20,14 +37,51 @@ const ROUNDS_UNTIL_SLEEPY: usize = 32;
...
+    }
+
+    #[inline]
+    pub fn sub_external_waiter(&self) {
+        let mut data = self.data.lock().unwrap();
+        data.external_waiters -= 1;
+    }
+
+    #[inline]
+    pub fn block(&self, deadlock_handler: &Option<Box<DeadlockHandler>>) {
```
Comment (michaelwoerister) [inline rayon-core/src/sleep/mod.rs]: I would prefer this to be called something like `mark_blocked()`.

File: rayon-core/src/sleep/mod.rs [hunk 2]
```diff
@@ -20,14 +37,51 @@ const ROUNDS_UNTIL_SLEEPY: usize = 32;
...
+        debug_assert!(data.blocked_threads < data.worker_count);
+        debug_assert!(data.active_threads > 0);
+        data.active_threads -= 1;
+        data.blocked_threads += 1;
+
+        data.deadlock_check(deadlock_handler);
+    }
+
+    #[inline]
+    pub fn unblock(&self) {
```
Comment (michaelwoerister) [inline rayon-core/src/sleep/mod.rs]: Same here, `mark_unblocked()` seems clearer to me.

File: rayon-core/src/registry.rs [hunk 3]
```diff
@@ -412,6 +425,24 @@ impl Registry {
     }
 }
 
+/// Mark a Rayon worker thread as blocked. This triggers the deadlock handler
```
Comment (nikomatsakis) [inline rayon-core/src/registry.rs]: Hmm. This seems ok — I was thinking that we might also just have some kind of callback whenever a thread goes to sleep. Then we can (on the rustc side) maintain the "blocked vs unblocked" accounting, and not complicate the Rayon side of the API.

File: rayon-core/src/sleep/mod.rs [hunk 4]
```diff
@@ -20,14 +37,51 @@ const ROUNDS_UNTIL_SLEEPY: usize = 32;
...
+                active_threads: worker_count,
+                blocked_threads: 0,
+                external_waiters: 0,
+            }),
             tickle: Condvar::new(),
         }
     }
 
+    #[inline]
+    pub fn add_external_waiter(&self) {
```
Comment (nikomatsakis) [inline rayon-core/src/sleep/mod.rs]: I'd like to see a comment on what an "external waiter" is

File: rayon-core/src/sleep/mod.rs [hunk 5]
```diff
@@ -1,15 +1,32 @@
 //! Code that decides when workers should go to sleep. See README.md
 //! for an overview.
 
+use DeadlockHandler;
```
Comment (nikomatsakis) [inline rayon-core/src/sleep/mod.rs:4]: We should document the changes here in [the README.md file for this module](https://github.com/Zoxc/rayon/blob/f541d2fa75cad1731d13d0e84095fcf2f05bef32/rayon-core/src/sleep/README.md)

Comment (nikomatsakis) [reply rayon-core/src/sleep/mod.rs:4]: It seems like it could just go in a little section at the end

File: rayon-core/src/registry.rs [hunk 6]
```diff
@@ -354,6 +356,17 @@ impl Registry {
         let job = StackJob::new(|injected| {
             let worker_thread = WorkerThread::current();
             assert!(injected && !worker_thread.is_null());
+            let worker_thread = &*worker_thread;
+            worker_thread.registry.sleep.add_external_waiter();
```
Comment (nikomatsakis) [inline rayon-core/src/registry.rs]: I don't really understand what an "external waiter" is

Comment (nikomatsakis) [reply rayon-core/src/registry.rs]: Oh, I ... guess I maybe do. It's just some external thread that is blocked waiting on Rayon to finish up?

Comment (nikomatsakis) [reply rayon-core/src/registry.rs]: What is the reason to have this? Like, wouldn't we want to be called back regardless if the worker threads have deadlocked?

**Review by nikomatsakis — APPROVED** [review]
This looks good, but what do you think about simplifying the inferface?

What I have mind is this:

We always get a callback whenever threads go to sleep. Then the "mark-blocked" logic (and, I suppose, the "external waiter" logic) can move into rustc — with each callback, we track the number of active threads, and if we ever find that the number of active threads equals the number of blocks threads, we recover somehow.

The disadvantage I see to this is that we will get more callbacks. The advantage is that the Rayon API is quite a bit simpler (just a callback).

File: rayon-core/src/registry.rs (hunk 6 above)
Comment (Zoxc) [reply rayon-core/src/registry.rs]: If we did not have this, we'd immediately detect a deadlock when a thread pool is created, as all the workers are idle.

**Comment by Zoxc** [comment]
> This looks good, but what do you think about simplifying the inferface?

That doesn't seem beneficial to me. It leaks more internal details about Rayon (how many external waiter there are, when thread go to sleep), reduces performance due to callbacks. You wouldn't be able to use the mutex in `Sleep`, which means you'd need an additional mutex on the rustc side. I don't find more mutexes helpful for reasoning about code.

File: rayon-core/src/sleep/mod.rs (hunk 1 above)
Comment (Zoxc) [reply rayon-core/src/sleep/mod.rs]: Renamed.

File: rayon-core/src/sleep/mod.rs (hunk 2 above)
Comment (Zoxc) [reply rayon-core/src/sleep/mod.rs]: Renamed.

File: rayon-core/src/sleep/mod.rs (hunk 5 above)
Comment (Zoxc) [reply rayon-core/src/sleep/mod.rs:4]: I added a section on deadlock detection: https://github.com/Zoxc/rayon/blob/rustc/rayon-core/src/sleep/README.md#deadlock-detection

File: rayon-core/src/sleep/mod.rs (hunk 4 above)
Comment (Zoxc) [reply rayon-core/src/sleep/mod.rs]: Added.

File: rayon-core/src/registry.rs (hunk 6 above)
Comment (nikomatsakis) [reply rayon-core/src/registry.rs]: As I wrote on IRC, I am wondering now if we can remove the concept of "external waiter", and instead just only issue a callback when there is at least one blocked thread.

Comment (Zoxc) [reply rayon-core/src/registry.rs]: I've removed the external waiter counter.

**Review by nikomatsakis — APPROVED** [review]
Looks good!

**Comment by nikomatsakis** [comment]
I'm happy with this new interface — it seems "sufficiently minimal". That is, informing Rayon when you are about to block the thread seems pretty reasonable, and getting a callback when there is no work other than your blocked threads seems good. It occurs to me that calling this *deadlock* is perhaps overstating the case -- it might be of course that one of your threads is blocked on I/O or something. But we can leave the name for now.

**Pull request closed without merging**